src/coreclr/src/jit/assertionprop.cpp

```diff
@@ -1959,12 +1958,48 @@ AssertionInfo Compiler::optAssertionGenJtrue(GenTree* tree)
     {
         std::swap(op1, op2);
     }
+
+    ValueNum op1VN = vnStore->VNConservativeNormalValue(op1->gtVNPair);
+    ValueNum op2VN = vnStore->VNConservativeNormalValue(op2->gtVNPair);
     // If op1 is lcl and op2 is const or lcl, create assertion.
     if ((op1->gtOper == GT_LCL_VAR) &&
         ((op2->OperKind() & GTK_CONST) || (op2->gtOper == GT_LCL_VAR))) // Fix for Dev10 851483
     {
         return optCreateJtrueAssertions(op1, op2, assertionKind);
     }
+    else if (vnStore->IsVNCheckedBound(op1VN) && vnStore->IsVNInt32Constant(op2VN))
+    {
```

Comment:
I believe that you also need a check for `(relop->gtOper == GT_EQ)` here, as it could also be `GT_NE`
Note that we may add additional relops above, such as `GT_LE` in the future.

Reply:
Below it appears that you are also trying to handle `GT_NE`
We will need test cases for both of these, with positive and negative tests.

Reply:
I added some comments to help clarify this, and an assert that the relop is one that we expect. There are test cases for both of those scenarios, let me know if you can think of others to add.